crates/matrix-sdk/src/oidc/mod.rs

```diff
@@ -244,6 +282,84 @@ impl Oidc {
         Self { client }
     }
 
+    fn ctx(&self) -> &OidcContext {
+        &self.client.inner.oidc_context
+    }
+
+    /// Sets the save/restore session callbacks for OIDC.
+    pub fn set_callbacks(
+        &self,
+        reload_session_callback: Box<ReloadSessionCallback>,
+        save_session_callback: Box<SaveSessionCallback>,
+    ) -> Result<(), OidcError> {
+        self.ctx()
+            .reload_session_callback
+            .set(reload_session_callback)
+            .map_err(|_| OidcError::DuplicateCallbacks)?;
+
+        self.ctx()
+            .save_session_callback
+            .set(save_session_callback)
+            .map_err(|_| OidcError::DuplicateCallbacks)?;
+
+        Ok(())
+    }
+
+    /// Enable a cross-process store lock on the state store, to coordinate
+    /// refreshes across different processes.
+    pub async fn enable_cross_process_refresh_lock(
+        &self,
+        lock_value: String,
+    ) -> Result<(), OidcError> {
+        // FIXME: it must be deferred only because we're using the crypto store and it's
+        // initialized only in `set_session_meta`, not if we use a dedicated
+        // store.
+        let mut lock = self.ctx().deferred_cross_process_lock_init.lock().await;
+        if lock.is_some() {
+            return Err(CrossProcessRefreshLockError::DuplicatedLock.into());
+        }
+        *lock = Some(lock_value);
+
+        Ok(())
+    }
+
+    /// Performs a deferred cross-process refresh-lock, if needs be, after an
+    /// olm machine has been initialized.
+    ///
+    /// Must be called after `set_session_meta`.
+    async fn deferred_enable_cross_process_refresh_lock(&self) -> Result<()> {
+        let deferred_init_lock = self.ctx().deferred_cross_process_lock_init.lock().await;
+        let Some(lock_value) = deferred_init_lock.as_ref() else {
+            return Ok(());
+        };
+
+        // If the lock has already been created, don't recreate it from scratch.
+        if let Some(prev_lock) = self.ctx().cross_process_token_refresh_manager.get() {
+            let prev_holder = prev_lock.lock_holder();
+            if prev_holder == lock_value {
+                return Ok(());
+            }
+            warn!("recreating cross-process store refresh token lock with a different holder value: prev was {prev_holder}, new is {lock_value}");
```

Comment:
Overly long line, maybe use event fields?

```suggestion
            warn!(
                prev_holder, new_holder = lock_value,
                "recreating cross-process store refresh token lock with a different holder value"
            );
```